### PR TITLE
Add motion controller with speed ramps and distance/angle PIDs

### PR DESCRIPTION
**Pull request by tictacmenthe**

### CODE REVIEW

File: highlevel/main.py [hunk 1]
```diff
@@ -138,7 +163,7 @@ async def _get_container(simulation: bool, stub_lidar: bool,
     if simulation or stub_socket_can:
         i.provide('motor_board_adapter', LoopbackSocketAdapter)
     else:
-        reader, writer = await tcp.get_reader_writer('localhost', 32000)
+        reader, writer = await tcp.get_reader_writer('192.168.1.120', 14000)
```
Comment (bonnetn) [inline highlevel/main.py:166]: maybe move that to config

File: highlevel/robot/controller/motion/position.py [hunk 2]
```diff
@@ -18,48 +36,66 @@ class PositionController:
...
-        self.probe.emit("encoder_left", tick_left)
-        self.probe.emit("encoder_right", tick_right)
+        self.position_left = tick_to_mm(
+            tick_left, self.configuration.encoder_ticks_per_revolution,
+            self.configuration.wheel_radius)
+        self.position_right = tick_to_mm(
+            tick_right, self.configuration.encoder_ticks_per_revolution,
+            self.configuration.wheel_radius)
+
+        # self.probe.emit("encoder_left", self.position_left)
```
Comment (bonnetn) [inline highlevel/robot/controller/motion/position.py:67]: why is it commented?

File: highlevel/robot/controller/motion/position.py [hunk 3]
```diff
@@ -5,11 +5,29 @@
...
 from highlevel.logger import LOGGER
 from highlevel.robot.entity.configuration import Configuration
-from highlevel.robot.entity.type import Millimeter
-from highlevel.util.odometry import OdometryFunc
+from highlevel.robot.entity.type import Millimeter, Radian, Tick, MillimeterPerSec
+from highlevel.util.filter.odometry import OdometryFunc
 from highlevel.util.probe import Probe
 
 
+def tick_to_mm(tick: Tick, ticks_per_revolution: Tick,
```
Comment (bonnetn) [inline highlevel/robot/controller/motion/position.py:13]: These 2 functions look like util functions

File: highlevel/robot/controller/motion/position.py [hunk 4]
```diff
@@ -18,48 +36,66 @@ class PositionController:
     """
     def __init__(self, odometry_function: OdometryFunc,
                  configuration: Configuration, probe: Probe):
+        self.odometry = odometry_function
         self.configuration = configuration
         self.probe = probe
-        self.odometry = odometry_function
 
+        self.distance_travelled: Millimeter = 0.0
```
Comment (bonnetn) [inline highlevel/robot/controller/motion/position.py:43]: why do you type hint only half of the fields?

fields such as `self.position_right_last` could use a type hint to specify their unit

File: highlevel/robot/controller/motion/position.py (hunk 3 above)
Comment (bonnetn) [inline highlevel/robot/controller/motion/position.py:13]: these 2 functions look like util functions that could be extracted in the util package

File: highlevel/robot/controller/motion/trajectory.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        LOGGER.get().info('trajectory_controller_move_to',
+                          current=pos_current,
+                          target=pos_target,
+                          reverse=reverse)
+
+        if (pos_current - pos_target).euclidean_norm() < 1:
+            return
+
+        direction = pos_target - pos_current
+        distance = direction.euclidean_norm()
```
Comment (bonnetn) [inline highlevel/robot/controller/motion/trajectory.py:41]: you can compute distance oncee and move the condition after..

File: highlevel/robot/controller/motion/trajectory.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        result = await self.motion_controller.rotate(delta_angle)
+        if result != MotionResult.OK:
+            raise RuntimeError("Move to point failed during rotation")
+
+        result = await self.motion_controller.translate(distance)
+        if result != MotionResult.OK:
+            raise RuntimeError("Move to point failed during translation")
+
+
+def normalize_angle(angle: float) -> float:
```
Comment (bonnetn) [inline highlevel/robot/controller/motion/trajectory.py:60]: Move to util?

File: highlevel/robot/controller/motion/trajectory_test.py [hunk 7]
```diff
@@ -0,0 +1,3 @@
+"""
+Test the trajectory module.
```
Comment (bonnetn) [inline highlevel/robot/controller/motion/trajectory_test.py:2]: is it really tho? :D

File: highlevel/robot/controller/strategy.py [hunk 8]
```diff
@@ -47,21 +50,71 @@
     (Vector2(300, 1450), True),
 ]
 
+PATH_MIRRORED = [(Vector2(pos.x, 2000 - pos.y), reverse)
+                 for pos, reverse in PATH]
+
+
+class ControlMode(Enum):
```
Comment (bonnetn) [inline highlevel/robot/controller/strategy.py:57]: Not sure what it has to do with the "strategy" of the robot?

same thing for `set_mode`

When opening strategy.py I would expect to read what is the very high level algorithm of the robot, not hardware/low-level specific code

File: highlevel/robot/controller/strategy.py [hunk 9]
```diff
@@ -47,21 +50,71 @@
...
+                self.configuration.pid_constants_position_right.k_d,
+            )
+            await self.trajectory_controller.motion_controller.motor_gateway.set_control_mode(
+                False, True)
 
     async def run(self) -> None:
         """
         Run the strategy.
         """
-        for vec, reverse in PATH:
```
Comment (bonnetn) [inline highlevel/robot/controller/strategy.py:62]: So, if we are removing this, why are we keeping PATH and PATH_MIRRORED?

I'd prefer to keep this

File: highlevel/robot/entity/configuration.py [hunk 10]
```diff
@@ -26,17 +28,44 @@ class Configuration:
...
+    # # Robot data
     robot_width: Millimeter
     robot_length: Millimeter
-
-    field_shape: Tuple[Millimeter, Millimeter]
-
     wheel_radius: Millimeter
     encoder_ticks_per_revolution: int
     distance_between_wheels: Millimeter
+    encoder_update_rate: int
```
Comment (bonnetn) [inline highlevel/robot/entity/configuration.py:46]: Use proper unit here please i.e. `Hz

small descriptions would also be appreciated

File: highlevel/robot/entity/type.py [hunk 11]
```diff
@@ -33,3 +35,14 @@ class Direction(Enum):
     BACKWARD = 'BACKWARD'
     LEFT = 'LEFT'
     RIGHT = 'RIGHT'
+
+
+class MotionResult(Enum):
```
Comment (bonnetn) [inline highlevel/robot/entity/type.py:40]: This seems very very specific to the motion control and no way generic to all the robot.

Move this to the controller file?

File: highlevel/robot/gateway/motor.py [hunk 12]
```diff
@@ -1,73 +1,156 @@
...
+            self, pid_constants: PIDConstants) -> PIDCoefficients:
+        """
+        Converts a PIDConstants structure to a Protobuf PIDCoefficients sub message.
+        Applies a scaling factor to each value.
+        """
+        scale_factor = self.configuration.pid_scale_factor
+        update_rate = self.configuration.motor_update_rate
+        max_value_on_motor_board = (2**32) - 1
+        if pid_constants.k_i * scale_factor / update_rate >= max_value_on_motor_board:
+            LOGGER.get().error('pid_to_proto_ki_too_large',
```
Comment (bonnetn) [inline highlevel/robot/gateway/motor.py:44]: raise an exception? It's probably dangerous to silently fail here

File: highlevel/robot/gateway/motor.py [hunk 13]
```diff
@@ -1,73 +1,156 @@
...
 class MotorGateway:
     """
     Motor gateway.
     """
-    def __init__(self, motor_board_adapter: SocketAdapter):
+    def __init__(self, motor_board_adapter: SocketAdapter,
+                 configuration: Configuration):
+        self.configuration = configuration
         self.motor_board_adapter = motor_board_adapter
+        self.pid_speed_left = PIDConstants(0, 0, 0)
```
Comment (bonnetn) [inline highlevel/robot/gateway/motor.py:22]: This gateway seems stateful, why is that?

Usually gateways are just... gateways to the hardware you should not expect it to hold any state, it provides functions to communicate with a lower level and that's it

File: highlevel/robot/gateway/motor.py [hunk 14]
```diff
@@ -1,73 +1,156 @@
...
+        Sets each wheel's PWM output. Ratios are duty cycles, signs are directions.
+        """
+        LOGGER.get().info('motor_gateway_set_pwms',
+                          ratio_left=ratio_left,
+                          ratio_right=ratio_right)
+        message = BusMessage(wheelPWM=WheelPWMMsg(ratio_left=ratio_left,
+                                                  ratio_right=ratio_right))
+        await self._send_message(message)
+
+    async def set_target_speeds(self, tick_left: TickPerSec,
```
Comment (bonnetn) [inline highlevel/robot/gateway/motor.py:81]: nit: rename tick_left -> speed_left

File: highlevel/robot/gateway/motor.py [hunk 15]
```diff
@@ -1,73 +1,156 @@
...
         message = BusMessage(moveWheelAtSpeed=MoveWheelAtSpeedMsg(
-            left_tick_per_sec=round(left_speed),
-            right_tick_per_sec=round(right_speed)))
+            left_tick_per_sec=round(tick_left),
+            right_tick_per_sec=round(tick_right)))
         await self._send_message(message)
 
-    async def send_pid(self, pid_left: PIDValues,
-                       pid_right: PIDValues) -> None:
+    async def set_target_positions(self, tick_left: Tick,
```
Comment (bonnetn) [inline highlevel/robot/gateway/motor.py:94]: do we still need that function?

File: highlevel/robot/gateway/motor.py [hunk 16]
```diff
@@ -1,73 +1,156 @@
...
+        self.pid_position_left = PIDConstants(left_kp, left_ki, left_kd)
+        self.pid_position_right = PIDConstants(right_kp, right_ki, right_kd)
+
+        LOGGER.get().info('motor_gateway_set_position_pids',
+                          pid_left=self.pid_position_left,
+                          pid_right=self.pid_position_right)
+        await self._send_pid_configs()
+
+    # pylint:disable=too-many-arguments
+    async def set_pid_speed(self, left_kp: float, left_ki: float,
```
Comment (bonnetn) [inline highlevel/robot/gateway/motor.py:122]: don't you have a PID entity btw?

File: highlevel/robot/gateway/motor_test.py [hunk 17]
```diff
@@ -14,12 +17,12 @@ class TestMotorGateway:
...
     @pytest.mark.asyncio
-    async def test_set_speed(socket_adapter_mock):
+    async def test_set_speed(socket_adapter_mock, configuration_test):
         """
-        Test that the motor gateway encodes a wheel speed message.
+        Test that the motor gateway encodes a wheel target speed message.
         """
-        motor_gateway = MotorGateway(motor_board_adapter=socket_adapter_mock)
-        await motor_gateway.set_speed(10, 20)
+        motor_gateway = MotorGateway(socket_adapter_mock, configuration_test)
```
Comment (bonnetn) [inline highlevel/robot/gateway/motor_test.py:24]: use KW values when calling constructors,

File: highlevel/robot/router.py [hunk 18]
```diff
@@ -43,6 +48,11 @@ async def dispatch_message(self, bus_message: BusMessage,
             LOGGER.get().info("low_level_log",
                               content=bus_message.debugLog.content,
                               source=source)
+        elif type_msg in [
```
Comment (bonnetn) [inline highlevel/robot/router.py:51]: at least add a comment on why they are ignored

also you can extract that list and convert it to a set for O(1) access

File: highlevel/simulation/controller/runner.py [hunk 19]
```diff
@@ -35,15 +39,26 @@ def __init__(self, event_queue: EventQueue,
...
         self.probe = probe
 
         self.tick = 0
         self.running = True
 
+        self.position_noise = 0  # percentage of noise, used as a scale for gaussian noise
+        self.position_delay = 50  # ticks before a position target is achieved
+
+        self.state.position_queue_left = deque(
+            [0 for _ in range(self.position_delay)])
```
Comment (bonnetn) [inline highlevel/simulation/controller/runner.py:52]: self.state.position_queue_left = deque([0]*self.position_delay)

File: highlevel/simulation/controller/runner_test.py [hunk 20]
```diff
@@ -38,48 +38,6 @@ def simulation_runner_factory(event_queue, simulation_gateway_mock,
...
-                     tick_offset=0)
-
-    task = asyncio.create_task(simulation_runner.run())
-    await asyncio.sleep(0.05)
-    task.cancel()
-
-    simulation_gateway_mock.movement_done.assert_called_once()
-
-
 @pytest.mark.asyncio
```
Comment (bonnetn) [inline highlevel/simulation/controller/runner_test.py:41]: no test for new logic?

File: highlevel/simulation/entity/event.py [hunk 21]
```diff
@@ -10,6 +10,7 @@ class EventType(Enum):
     """
     Event type.
     """
+    WHEEL_SPEED = 'WHEEL_SPEED'
     MOVE_WHEEL = 'MOVE_WHEEL'
     MOVEMENT_DONE = 'MOVEMENT_DONE'
```
Comment (bonnetn) [inline highlevel/simulation/entity/event.py:15]: movement done and move_wheels are gone, no?

remove?

File: highlevel/simulation/entity/simulation_state.py [hunk 22]
```diff
@@ -33,17 +28,9 @@ class SimulationState:
     cups: List[Cup]
     left_tick: int
     right_tick: int
+    position_queue_left: Deque[int]
```
Comment (bonnetn) [inline highlevel/simulation/entity/simulation_state.py:31]: maybe use unit here?

File: highlevel/simulation/router_test.py [hunk 23]
```diff
@@ -24,78 +23,15 @@ def simulation_router_setup(configuration_test, event_queue_mock,
...
-    """
-    If we receive a packet to move wheels by 0 unit, do nothing.
-    """
-    bus_message = BusMessage(translate=TranslateMsg(ticks=0))
     msg_bytes = bus_message.SerializeToString()
     await simulation_router.handle_movement_order(msg_bytes, 'test')
 
-    event_queue_mock.push.assert_called_with(
-        EventOrder(type=EventType.MOVEMENT_DONE, payload=None), 0)
+    assert simulation_state_mock.position_queue_left[-1] == 100
```
Comment (bonnetn) [inline highlevel/simulation/router_test.py:36]: no test for new logic?

File: highlevel/util/perf_metrics.py [hunk 24]
```diff
@@ -30,7 +30,7 @@ async def print_performance_metrics() -> None:
         active_tasks = len(tasks)
 
         vmem = psutil.virtual_memory()
-        LOGGER.get().info(
+        LOGGER.get().debug(
```
Comment (bonnetn) [inline highlevel/util/perf_metrics.py:33]: why? maybe increase the time interval if it bothers you?  
Also does not seem related to this diff

(

File: highlevel/util/probe.py [hunk 25]
```diff
@@ -33,15 +33,26 @@ class Probe:
     def __init__(self, clock: Clock):
         self._clock = clock
         self._event_log: List[DebugEvent] = []
+        self.count_left = 0
```
Comment (bonnetn) [inline highlevel/util/probe.py:36]: rm changes from that file

File: tool/script/read_bus.py [hunk 26]
```diff
@@ -2,15 +2,17 @@
...
 from google.protobuf import json_format
 
 from highlevel.adapter.socket.socket_adapter import TCPSocketAdapter
 from highlevel.util import tcp
 from proto.gen.python.outech_pb2 import BusMessage
 
+t_last = 0.0
 
 async def main():
+    global t_last
```
Comment (bonnetn) [inline tool/script/read_bus.py:15]: not sure whats going on there or why it's related to that commit

File: highlevel/Makefile [hunk 27]
```diff
@@ -34,3 +34,8 @@ run-simulation:
 .PHONY: run
 run:
 	PYTHONPATH=${CURDIR}/.. OUTECH_SIMULATION=false python -m highlevel.main
+
+.PHONY: run-stub
+run-stub:
```
[the remark this answers is not in the document]
Comment (tictacmenthe) [reply highlevel/Makefile:39]: Changed to ```run-lidarless```.

This is related to the PR as it is used to run it on hardware. We'll keep it.

File: highlevel/main.py [hunk 28]
```diff
@@ -53,20 +56,37 @@
...
+    pid_limits_angle=PIDLimits(4.0, 2, 0.000),
 )
 
 SIMULATION_CONFIG = SimulationConfiguration(
     speed_factor=1e100,  # Run the simulation as fast as possible.
-    tickrate=60,
-    rotation_speed=math.pi * 2 * 4.547,
-    encoder_position_rate=100,
-    replay_fps=60,
+    tickrate=1000,
```
[the remark this answers is not in the document]
Comment (tictacmenthe) [reply highlevel/main.py:87]: you chose it :1st_place_medal:

File: highlevel/main.py [hunk 29]
```diff
@@ -87,25 +107,29 @@ async def _get_container(simulation: bool, stub_lidar: bool,
...
     i.provide('obstacle_controller', ObstacleController)
     i.provide('debug_controller', DebugController)
     i.provide('match_action_controller', MatchActionController)
 
-    i.provide('motor_gateway', MotorGateway)
-
     i.provide('probe', Probe)
     i.provide('event_loop', asyncio.get_event_loop())
 
+    i.provide('simulation_configuration', SIMULATION_CONFIG)
```
[the remark this answers is not in the document]
Comment (tictacmenthe) [reply highlevel/main.py:127]: idk, you wrote that line :eyes: 
It does not really matter if it is provided or not in any case.

File: highlevel/robot/controller/motion/position.py (hunk 3 above)
Comment (tictacmenthe) [reply highlevel/robot/controller/motion/position.py:13]: Moved to robot.entity.type as they are related to the units defined there.

File: highlevel/robot/controller/motion/trajectory_test.py (hunk 7 above)
Comment (tictacmenthe) [reply highlevel/robot/controller/motion/trajectory_test.py:2]: The tests are left as en exercise to the reader.

File: highlevel/robot/controller/strategy.py (hunk 9 above)
Comment (tictacmenthe) [reply highlevel/robot/controller/strategy.py:62]: You can revert to it if needed.
There is no strategy in the robot anyway, should we remove the strategy controller too?

File: highlevel/robot/gateway/motor.py (hunk 12 above)
Comment (tictacmenthe) [reply highlevel/robot/gateway/motor.py:44]: It is supposed to log an error, that's not silent.

File: highlevel/robot/gateway/motor.py (hunk 14 above)
Comment (tictacmenthe) [reply highlevel/robot/gateway/motor.py:81]: renamed to left, right

File: highlevel/robot/gateway/motor.py (hunk 15 above)
Comment (tictacmenthe) [reply highlevel/robot/gateway/motor.py:94]: We will, when doing an action the robot needs to stay still.

File: highlevel/robot/gateway/motor.py (hunk 16 above)
Comment (tictacmenthe) [reply highlevel/robot/gateway/motor.py:122]: this is due to limitations on the debug interface made by our lead web dev.

File: highlevel/robot/gateway/motor_test.py (hunk 17 above)
Comment (tictacmenthe) [reply highlevel/robot/gateway/motor_test.py:24]: We would have to go through all the repo, as it is not done everywhere.
We can refactor that in a future clean.

File: highlevel/simulation/entity/event.py (hunk 21 above)
Comment (tictacmenthe) [reply highlevel/simulation/entity/event.py:15]: not really related, we will have to see later if the event queue is even needed anymore.

File: highlevel/main.py (hunk 28 above)
[the remark this answers is not in the document]
Comment (bonnetn) [reply highlevel/main.py:87]: let's put 100 here, 1000 might slow down the simulation too much

File: highlevel/main.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (bonnetn) [reply highlevel/main.py:126]: it does not really make sense to provide simulation dependencies when running in production...
I'd prefer to avoid having simulation stuff running at the cup

File: highlevel/robot/controller/motion/trajectory.py (hunk 5 above)
Comment (bonnetn) [reply highlevel/robot/controller/motion/trajectory.py:37]: did you remove the condition alltogether? why?

File: highlevel/robot/controller/strategy.py (hunk 9 above)
Comment (bonnetn) [reply highlevel/robot/controller/strategy.py:62]: There is a stub strategy (this long list of positions).
It's just very rough

I like the fact that I can use this long running scenario for testing

File: highlevel/robot/entity/type.py [hunk 30]
```diff
@@ -33,3 +36,21 @@ class Direction(Enum):
     BACKWARD = 'BACKWARD'
     LEFT = 'LEFT'
     RIGHT = 'RIGHT'
+
+
+def tick_to_mm(tick: Tick, ticks_per_revolution: Tick,
```
Comment (bonnetn) [inline highlevel/robot/entity/type.py:41]: Maybe we should move this "type" file out of entities to util...

File: highlevel/robot/gateway/motor.py (hunk 12 above)
Comment (bonnetn) [reply highlevel/robot/gateway/motor.py:51]: Ok but why not interrupt the flow of execution?

Is that an error that is expected to happen in real conditions?

File: highlevel/robot/gateway/motor.py (hunk 15 above)
Comment (bonnetn) [reply highlevel/robot/gateway/motor.py:101]: what's the difference bw using this function and set_speed(0,0) when we need the robot to stay still?

File: highlevel/robot/gateway/motor.py (hunk 16 above)
Comment (bonnetn) [reply highlevel/robot/gateway/motor.py:131]: hahahaha fair

File: highlevel/util/geometry/trigonometry.py [hunk 31]
```diff
@@ -0,0 +1,20 @@
+"""
+Trigonometry module
+"""
+import math
+
+from highlevel.robot.entity.type import Radian
+
+
+def normalize_angle(angle: Radian) -> Radian:
```
Comment (bonnetn) [inline highlevel/util/geometry/trigonometry.py:9]: a simple unit test would be nice

File: highlevel/robot/controller/motion/trajectory_test.py (hunk 7 above)
Comment (bonnetn) [reply highlevel/robot/controller/motion/trajectory_test.py:2]: no!

File: highlevel/robot/gateway/motor_test.py (hunk 17 above)
Comment (bonnetn) [reply highlevel/robot/gateway/motor_test.py:24]: We usually use KW arguments for calling constructors.

Anyway, here you are explicitely **removing** the kw arguments, please do not do that, that's a regression

File: highlevel/simulation/entity/event.py (hunk 21 above)
Comment (bonnetn) [reply highlevel/simulation/entity/event.py:15]: Mh... maybe we can remove the queue then, since this is tied to the arch refactoring that you are doing...

File: highlevel/robot/controller/motion/trajectory.py (hunk 5 above)
Comment (tictacmenthe) [reply highlevel/robot/controller/motion/trajectory.py:37]: put it back, as suggested.

File: highlevel/robot/entity/type.py (hunk 30 above)
Comment (tictacmenthe) [reply highlevel/robot/entity/type.py:41]: Agreed, I was thinking about doing it.

**Comment by bonnetn** [comment]
**looks fine..**

https://twitter.com/iamdevloper/status/397664295875805184?lang=en
image
